src/lib/fields_derivers_graphql/fields_derivers_graphql.ml

```diff
@@ -0,0 +1,335 @@
+              -> ('ctx, 'output_type option) Schema.typ
+          }
+      end
+
+      module Accumulator = struct
+        module Elem = struct
+          type 'input_type t =
+            { run : 'ctx. unit -> ('ctx, 'input_type) Schema.field }
+        end
+
+        (** thunks generating the schema in reverse *)
+        type 'input_type t = 'input_type Elem.t list
+      end
+
+      let init () = []
+
+      let add_field (type f input_type) :
+             f Input.t
+          -> ( [< `Read | `Set_and_create ]
+             , input_type
+             , f )
+             Fieldslib.Field.t_with_perm
+          -> input_type Accumulator.t
+          -> (input_type Creator.t -> f) * input_type Accumulator.t =
+       fun t_field field acc ->
+        ( (fun _ -> failwith "Unused")
+        , { Accumulator.Elem.run =
+              (fun () ->
+                Schema.field
+                  (Fields_derivers.name_under_to_camel field)
+                  ~args:Schema.Arg.[]
+                  ?doc:None ?deprecated:None ~typ:(t_field.run ())
+                  ~resolve:(fun _ x -> Field.get field x))
+          }
+          :: acc )
+
+      (* TODO: Do we need doc and deprecated and name on finish? *)
+      let finish ((_creator, schema_rev_thunk) : 'u * 'input_type Accumulator.t)
+          : 'input_type Output.t =
+        { Output.run =
+            (fun ?doc ~name ->
+              Schema.obj name ?doc ~fields:(fun _ ->
+                  List.rev
+                  @@ List.map schema_rev_thunk ~f:(fun f ->
+                         f.Accumulator.Elem.run ())))
+        }
+
+      let int_opt_ = Input.{ run = (fun ?doc:_ () -> Schema.int) }
+
+      let int_ = Input.{ run = (fun ?doc:_ () -> Schema.(non_null int)) }
+
+      let string_opt_ = Input.{ run = (fun ?doc:_ () -> Schema.string) }
+
+      let string_ = Input.{ run = (fun ?doc:_ () -> Schema.(non_null string)) }
+
+      let bool_opt_ = Input.{ run = (fun ?doc:_ () -> Schema.bool) }
+
+      let bool_ = Input.{ run = (fun ?doc:_ () -> Schema.(non_null bool)) }
+
+      let list_ (x : 'a Input.t) =
+        Input.{ run = (fun ?doc:_ () -> Schema.(non_null (list (x.run ())))) }
+    end
+
+    include Fields_derivers.Make (Deriver_basic)
+  end
+end
+
+module Graphql_fields = Graphql_fields_raw.Make (struct
+  include Async_kernel.Deferred
+
+  let bind x f = bind x ~f
+
+  module Stream = struct
+    type 'a t = 'a Async_kernel.Pipe.Reader.t
+
+    let map t f =
+      Async_kernel.Pipe.map' t ~f:(fun q ->
+          Async_kernel.Deferred.Queue.map q ~f)
+
+    let iter t f = Async_kernel.Pipe.iter t ~f
+
+    let close = Async_kernel.Pipe.close_read
+  end
+end)
+
+(** Convert this async Graphql_fields schema type into the official
+    Graphql_async one. The real Graphql_async functor application *is*
+    equivalent but the way the library is designed we can't actually see it so
+    this boils down to an Obj.magic. *)
+let typ_conv (typ : ('a, 'b) Graphql_fields.Schema.typ) :
```

Comment:
Consume the schema as a functor arg, perhaps? `Graphql_fields_raw.Raw` doesn't really own the schema anyway, so it seems strange for the functor to instantiate it.

Reply:
It's possible we can avoid this when we hook it up to the real graphql server actually if we just switch the other graphql code to use this instatiation of the functor. Future work